### DISCUSSION TITLE
Versatile Thermostat would be still more versatile with a PID controller.

### DISCUSSION THREAD

**Discussion by peterniebert** [jmcollin78/versatile_thermostat#184] [Ideas]
First of all, really nice dialogues and features!

I am experimenting this control on an appartment with electrical heaters. The problem I have with it is that there are too many difficult to model influences and that TPI does not solve overshooting issues. For instance, the appartment I look at has windows in the living room that sometimes receive sun shine that significantly contributes to the heating. You can observe that this is happening by a rapidly rising temperature, but the stateless TPI controller cannot anticipate the overshoot and continues as though the temperature where constant.

There is an interesting PID controller integration in HACS, which however is otherwise limited:
https://github.com/ScratMan/HASmartThermostat

So it would be nice to add PID as an alternative control algorithm. I do not have the time right now to do it, I was wondering whether this is on your todo list.

Kind regards and thank you for your nice work!
Peter

**Comment by maia** [comment]
Peter, have you tried the "Self-regulation" that's available in version 3.8? While it does not yet help in all situations, it's a good start. And I'd welcome a discussion about different algorithms and especially the combination of different timeframes. For example I'm surrounded by 50cm brick walls, so heating with a gas heater and radiators is very, very slow. Other's have electrical heaters and can heat a room by 3°C within 15 minutes. This also has an influence on what happenes when one has the window open for 10 minutes or when the room is exposed to solar energy. So even if you don't have time to implement it, when you can come up with good suggestions, someone else might be able to contribute to the feature.

**Comment by jmcollin78** [comment]
One of the first issue I create was this one #8 .
But finally the TPI algorithm works really fine for VTherm over_switch or over_valve.
And for the VTherm over_climate, there is now an auto-regulation feature which works also fine for many people as me.

> https://github.com/ScratMan/HASmartThermostat

Yes, I was in contact with him. His feedback was: it is a nightmare to tune and the auto-tuning doesn't work. So for the moment, this enhancement is not planed. I don't have enough feedback on auto-regulation to start implementing a new algorithm which will add another stack of complexity.

**Comment by jmcollin78** [comment]
> I am experimenting this control on an appartment with electrical heaters. The problem I have with it is that there are too many difficult to model influences and that TPI does not solve overshooting issues. 

The TPI was designed for electrical heater. It should do the job !. Maybe you just need a little tuning of the default coef_int and coef_ext values.
Can you please send some graphs wich shows the problem and I think I will be able to help.

You can also check here to start tuning:
https://github.com/jmcollin78/versatile_thermostat/blob/main/README.md#examples-tuning 
https://github.com/jmcollin78/versatile_thermostat/blob/main/README.md#tpi-algorithm

**Comment by peterniebert** [comment]
Thanks for all of your reactions! I have a hard time recovering graphs of overshoots as I have modified the parameters to avoid them, but with the current parameters, you can still see some dynamics.

Attached is a graph over the last 60h of the concerned room. On top, the external temperature as communicated by Meteo France, in the middle, the target temperature and the measured temperature of the thermostat. Bottom, the power percentage as assigned by TPI. I use 15 minute periods and the thermometer typically reports between 10 and 30 minutes depending on the change rate. I have already worked quite a bit on the parameters of TPI prior to this graph.

As you can see, coeff_ext may be slightly on the lower end as the temperature remains most of the time below the graph. This is a choice I made because otherwise I really get overshoots of up to 0.5 degrees since the external temperature is just one factor and does not capture all that is changing in the environment.

There is an artefact around 11h on 11/11 as we added an auxiliary heater to the room to increase its reactivity and I tried to see the reaction by short manipulations of the target temperature, but the overshoot following is sunshine falling onto the room. 

![graph](https://github.com/jmcollin78/versatile_thermostat/assets/34269983/9304ed76-09c7-4768-be75-ee01e0ea3cbd)

My interest/hope in PID is really on the D part which could allow to anticipate dynamically. For instance, it could have permitted to anticipate the undershoot at 11/11 15h, when the temperature was still above 19.5 but steeply falling.

By the way, I noticed that with the electric heaters sitting on the walls, the overall consumption increases when the temperature is allowed to temporarily drop as in this graph. My explanation would be increased losses to the walls when the heaters have to catch up and go to a higher duty cycle. Has anyone made similar observations?

**Reply by maia** [reply]
Peter, could you add the entity for the regulated temperature to the graphs? Something similar to this here? It's helpful to see all temperatures in a single graph: target temperature, room temperature, regulated temperature and (if available) temperature measured at the heating device.
<img width="1167" alt="Bildschirmfoto 2023-11-12 um 09 22 05" src="https://github.com/jmcollin78/versatile_thermostat/assets/9836/077560aa-da89-4d52-97b0-521483e2bae1">

**Reply by maia** [reply]
Peter, once the cycle issue and the temperature slope issues are solved it should be possible to tell which direction the room temperature has been taking over the past x minutes and to predict the room temperature at the next cycle. And if the predicted temp is above the target, the heater could be turned down preemtively.

**Reply by peterniebert** [reply]
There already are target temperature and measured temperature (in the middle), I do not have access to the temperature of the heaters which run behind switches doing on-off in a stupid mode. On the heaters themselves, the temperature is set to something like 28, so that when Versatile Thermostat says "on", they run at maximum power until switched off.

**Reply by maia** [reply]
If you‘re using the self-regulated mode you should have an entity named „regulated temperature“ for the virtual_thermostat service, do you have that and if yes can you show that too?

**Reply by peterniebert** [reply]
Sorry, I did not get this right, I guess: I am using Versatile Thermostat in mode "thermostat on a switch", not on another thermostat. Are you suggesting to cascade, i.e. create another instance of Versatile Thermostat to control the former with some form of selfregulation (which sounds like the I component in the PID model)?

**Reply by maia** [reply]
Oh, I’m sorry, I‘m not familiar with the „over switch“ mode and if the self-regulation exists in this mode.

**Reply by jmcollin78** [reply]
Thank you @maia  for trying to help me in the support activity. I will have a look because VTherm over switch doesn't have (or doesn't need ?) self-regulation.

**Comment by jmcollin78** [comment]
Hello @peterniebert ,

Some comments about your graphs:

1.
![Capture d’écran 2023-11-12 à 10 50 45](https://github.com/jmcollin78/versatile_thermostat/assets/1712374/db64f5bb-44c0-4cb7-bcb0-28658cd07b1e)

The temperature growth is very, very slow. With 100° heating, it takes from lets say 5:00 to 14:00 to gain 1°C. In such configuration, you should avoid to change the target temperature too much else the room temperature is never stable.

Maybe you should consider having a more powerful heater in this room or review isolation.

2.
![Capture d’écran 2023-11-12 à 10 57 08](https://github.com/jmcollin78/versatile_thermostat/assets/1712374/5107d686-21b3-4bfb-8843-e7c580453df6)

The overshoot is not a problem because the power is 0 during this phase. So there is no electricity loss in that case.

3. there is a stable and well regulated period here:
![Capture d’écran 2023-11-12 à 11 00 18](https://github.com/jmcollin78/versatile_thermostat/assets/1712374/1e7b7d3c-d2b0-4a23-9415-6d7f54f44fa2)

The power percent is around 40% and temperature is relatively stable around the target. This means your parameter are corrects but the heater is too low (40% is too much in a stable period. You should have 20% not more than this).

4. I agree with you that PID should help in that period of time. The temperature doesn't reach the target:
![Capture d’écran 2023-11-12 à 11 05 16](https://github.com/jmcollin78/versatile_thermostat/assets/1712374/f0252014-2274-45c1-b571-7f6949f8c29c)

But, your main problem is the under sizing of your heater. The power is already at 60% and it can't reach the target. Forcing to 80% with self-reguation may help but in terms of electrical consumption this is not very good. Change your heater to a more powerful and won't need any PID algorithm :-)

Here is a stable situation with a correctly sized heater:

The power is stable to 20% - 30%, and temperature is very near the target (+/- 0.1°C)
![Capture d’écran 2023-11-12 à 11 12 47](https://github.com/jmcollin78/versatile_thermostat/assets/1712374/5eda4a5d-d6d7-4038-a404-9d4e1a37d0a6)

**Reply by peterniebert** [reply]
Thanks, I agree with your observations that the heater is too slow. Maybe it is not as such underpowered, but I suspect it to produce heavy losses through the wall where it is fixed, in particular when it is working at its limit. Smaller heaters in the bed rooms, similarly powered when relating to the room surface, produce much snappier reactions.

It is one of many appartments built in the Paris outskirts in the 1980s, depending fully on electric heating (the nuclear power plants where very new and maybe underused then) and with poor insulation ... in a time where constructions in other countries already did much better.

**Comment by peterniebert** [comment]
I now understood what is going on, the limit of TPI modeling.

We substituted the wall mounted heater in the room shown above by an auxiliary electric heater in the "middle" of the room ... and got much better reactivity and 20% savings! So what is happening:

The TPI model supposes to the  that the thermal flow behaves much like a linear function of the temperature difference between inside and outside.  But when a heater (electric, in my case) mounted to an external wall goes to a higher level of operation, it also heats the wall behind it to a higher temperature (than the room temperature), meaning that the thermal flow from the inside to the outside increases as you heat! In other words, there is a quadratic(? have not tried to write equations) component in the function whose influence depends on the position of the heater itself.

The following graph shows a room with a wall mounted heater. This room is heated to 18 degrees most of the day and sometimes to 19, where an adjacent room is heated to 19 most of the day. As you can see, the external coefficient is too low for the higher target (which is not reached)  and too high for the lower target (not reached either) when the room is mostly heated by the rest of the appartment.

![tpi](https://github.com/jmcollin78/versatile_thermostat/assets/34269983/b97368b3-39d0-4b7d-a0ca-1cade0e38609)


Conclusion: it is an extremely common and extremely bad practice to place heaters on outside walls. And TPI poorly models such a bad practice, and this explains my initial problem.

**Reply by jmcollin78** [reply]
I wonder if I will not add a "I" factor in the over_switch regulation. The I factor ois the integral of the error which leads to a (I hope so) better self-regulation. I have implements this in over_climate VTherm and it gives great result for most people.

I a previous post you talk about the "D" factor. Can you tell me more on this, how to calculate it and how to tune it ?

**Comment by bontiv** [comment]
Hello,

I thinks it will be interesting to reopen this feature. Another thermostat has just implemented PID controller in their next release : https://github.com/KartoffelToby/better_thermostat

Maybe their work can be adapted to be implemented in VTherm?

**Reply by caiusseverus** [reply]
A PID controller is perhaps more suited to radiators and higher inertia systems because the integral term allows it to automatically correct a sustained error.

The problem is that radiators don't necessarily respond linearly to a 0 to 100% power level like an electric heater and that means an algorithm like the one in Vtherm where the static state is fixed won't necessarily be adequate for all cases.

I don't think the current algorithm uses an integral term at all for valve control so if the parameters are off it won't actively correct for that. The auto-tpi in development kind of helps but I'm finding that it's difficult to get to values which work well.

**Reply by jmcollin78** [reply]
Ping @KipK  😉  🤭

**Reply by KipK** [reply]
We're working on something related. But it's still highly early stage, and considered as an experiment.

https://github.com/KipK/versatile_thermostat/blob/PI/documentation/en/feature-autopi.md

The idea is to use thermal modeling to adjust dynamically Kp Ki gains on a PI loop.
No fixed coefficients.

Results are encouraging: https://drive.google.com/file/d/10001FEFbx5hbWJ-T6OerAvzSfAntNii5/view?usp=drivesdk

I'm just implementing @gael1980 work. That's too much math to me to add much value on this.

**Reply by jmcollin78** [reply]
For now we focus on the auto-TPI.

**Reply by caiusseverus** [reply]
That's going to be interesting. I will look forward to testing it.

**Reply by tbrasser** [reply]
In addition to all above I'd like to describe my system and "ideal" way to control it:

- combiboiler (in future will be heatpump) which is driven by "Smart Autotune Thermostat" based on room error (so it has all vtherm climates configured) and based on outside temperature, by setting the water flow temperature and the max (relative) modulation over opentherm (esphome). 

- trv on each radiator (8 in total), I know the flow and return water temperature(s), and using each radiator its design rated power, we can calculate the radiator effective maximum power (e.g. if valve were 100% open).

(delta_t / 50)^1.3

50=design delta t (between mean water and air temperature)
1.3=factor to raise to for radiators (nonlinear relationship)


I guess in short: take into account water temperature for valve control (by using desired power / effective max rad power), this would result in "normally open" (when water is cold) valves

@KipK Maybe doable once one has room thermal behaviour determined using auto-tpi?

**Reply by bontiv** [reply]
@KipK The PI algorithm is suffisent for me. I've hydrolic heater on my home. I'm very entousiasm about this new algorithm. A very big thanks for your work!

**Reply by bontiv** [reply]
I think I'll finally work on my end for a PID controller with MPC tuning by [Gekko](https://gekko.readthedocs.io/en/latest/). I'm curious about what can be done. I'll give a feedback at the end of testing my lab.

**Reply by KipK** [reply]
> I think I'll finally work on my end for a PID controller with MPC tuning by [Gekko](https://gekko.readthedocs.io/en/latest/). I'm curious about what can be done. I'll give a feedback at the end of testing my lab.

@bontiv  It's not far from what we're doing on SmartPI ( renamed autoPI to SmartPI to limit confusion with autoTPI. ) 
If you want to participate bump me on ha community or french ha forum. You probably have some values to add

**Reply by bontiv** [reply]
Don’t worry, it’s just a lab in private repository to test other models for system with low inertia (like electric heaters or PAC). It’s just a trash code, not destinated to be merged.

I think nobody want a big dependency on another python module nor his system do a contiuous machine learning which spend there resources on background.

**Reply by bontiv** [reply]
My reference document for advanced PID learning : [Tuning of PID Controllers Using Reinforcement Learning for Nonlinear System Control](https://www.mdpi.com/3209182), published by the Department of Automatic Control and Electronics, University of Craiova, 200585 Craiova, Romania.

**Reply by KipK** [reply]
Thanks.
Adding this ref too as a bible: https://www.ucg.ac.me/skladiste/blog_2146/objava_92847/fajlovi/Astrom.pdf

We're not doing machine learning on SmartPI. Trying to keep the algo light in ressources and dependencies.
 It's use AB based thermal model. Once A and B are trustfull enough, it then calculate Kp Ki, FF.

**Reply by bontiv** [reply]
Yes, I read many articles with the relay method of Astrom-Hagglund. They have done a very big work on PID. One of the last PID learning algorithm.

**Reply by KipK** [reply]
Seems your help could be precious; Don't you want to give a hand ?

**Reply by bontiv** [reply]
You work too fast for me 😄 I might just slow you.

I've not enough time to develop softwares. My last real developer experience was 13 yers ago. I developed the monitoring system (SCADA) of one datacenter of the TDF company (télédiffusion française). Since that experience, I was hired as manager of a production and operation service. I don't practice regularly (only devops), I'm not totally updated with lasts standards and I'm not as fast as this period.

I'll ask my CEO to get more time to work on open source projets. Maybe if it can reduce costs for my company, he will be agree to let me work on it.

**Reply by KipK** [reply]
That's not on coding we need help, I do this part. That's mainly for theory and maths behind the algorithm. 
 We're testing different things, going back and forth. 
Gael is alone on this part and  I can't follow , it's out of my league, just coding. :) 

Having an external view would help I think.

**Reply by bontiv** [reply]
No problem, I finish my little ML-PID lab and then I test your smartPI 😄

**Reply by KipK** [reply]
Ho it's not ready. There's still a bunch of work to do.
But definitively testable

**Reply by bontiv** [reply]
That's actually crazy. I finished the base of the PID-ML, I've 8 Gb of dependencies in my venv. With nvidia cuda, torch, gymnasium, etc.... Neural network take a lot of resources...

I continue with this machine learning for the PoC, but it's definitely impossible to merge it.

I also finished the base PID algorithm with two windup protections and derivation filter. It was not easy to add a new algorithm. There are methods which manipulate privates property (starting by "_") of another class and some processes which are not in the best place.

For example, the `on_time_sec` in the `prop_algo_xxx`. So all algo need the same piece of code which transform a on_percent into on_time_sec / off_time_sec.

It would be more easy if all `prop_algo_xxx` manage only the math of the algorithm (get a `calculated_on_percent` for a situation) and the `prop_handler_xxx` the link between the math and VTherm logic (on_time_sec / safety / on_percent / etc...).

But, I really love my PoC and when I work on it, I better understand how the project works and help / contribute.

**Reply by KipK** [reply]
Yep you're right. I've identified these architectural issues missing from the previous refactor and currently have a wip for exactly this:

- Improve encapsulation by avoiding direct access to private attributes.
- Centralize the duty-cycle timing logic to keep algorithms focused solely on the calculation math. 

I have not submitted the PR yet.

**Reply by bontiv** [reply]
@KipK je vais commencer à regarder de plus près SmartPI la semaine prochaine. Je ne sais pas comment tu veux que je fasse des retours dessus. Au cas où,  mon pseudo discord est bontiv.

**Reply by KipK** [reply]
je viendrais chat sur discord te dire où on en est. On attaque l'apprentissage du deadtime ( methode Sundaresan & Krishnaswamy )

**Reply by KipK** [reply]
@bontiv I've invited you on our discord server. 

A bit of reading first: 
in french: https://gist.github.com/KipK/1bb90a9594e1201216a4df3f0402e062
in english: https://gist.github.com/KipK/b2eef7d274806c80b88a4a7e199135af

**Reply by maia** [reply]
For these who might otherwise miss my comment on the gist linked by @kipk above:

This looks really promising and feels like a substantial improvement. I am very much looking forward to testing it and following its further development.

While reading the concept, I noticed one piece of information that, as far as I understand, is not currently used, but could be quite helpful in certain setups.

I see a potential issue in some installation configurations, for example in my own. I have twelve radiators equipped with Sonoff TRVZBs, a central boiler with an anti cycling lockout, and the boiler is controlled only by a single reference room. This means the boiler starts only when that reference room has a heating demand and the lockout period has elapsed. In other rooms, valves may already be open while no heat is actually being delivered yet, sometimes for many minutes. In my case this is barely noticeable in daily operation, because the building is very massive and has a high effective thermal mass, so the rooms are globally very slow to react. On top of that, my battery powered Zigbee temperature sensors only report changes from 0.2°C upwards and at most every five minutes. Since they are deliberately not mounted close to the radiators, they additionally measure a rather slow room temperature that is strongly influenced by the building fabric.

My concern is that, in such a setup, there is conceptually a mismatch between heating demand and effective heating output. There are phases where a demand is active and valves are open, but the heating power is not yet physically effective. This delay does not originate from the room physics, but from the heat generator logic and the sensor behavior. From the algorithm’s point of view, this appears as long periods without a measurable response, followed by sudden updates. In such situations, this phase can easily look like increased room inertia, even though it actually says nothing about the real thermal dynamics of the room.

This is also where the role of the measurement signal becomes conceptually important. Takeoff detection is very precise when the measured quantity reacts quickly and with sufficient resolution. With slow and quantized sensors, however, takeoff primarily marks the beginning of a detectable reaction in the measurement signal, not necessarily the onset of the physical heating effect itself. This does not question the idea of the approach, but it highlights that, in such setups, measurement logic and system logic can weigh more heavily in the interpretation than the actual room dynamics.

If, however, one were to use a piece of information that is already available in many setups, namely the temperature at the thermostat valve or radiator, this risk might be mitigated. Not as an additional control variable, but purely as an indicator of whether heating power is physically being delivered.

SmartPI could then distinguish more clearly between demand active and heat available. The measured indoor temperature would remain the control variable. The TRV temperature would merely serve as a gating signal, limiting learning and interpretation of the dynamics to phases in which the control output is physically effective. Phases with open valves but a blocked boiler would no longer be interpreted as properties of the room.

From my perspective, this would have two conceptual advantages. First, room parameters and delays would be less distorted by system logic and measurement behavior. Second, such distortions would also propagate less into derived terms like feed forward, without requiring any change to the control law itself. The core principle would remain unchanged. It would simply provide a clearer classification of the phases in which learning is meaningful.

What do you think?

**Reply by tbrasser** [reply]
@maia this is why I believe central heat source and emitter control cannot be looked at seperately. I would highly suggest the developers of Versatile Thermostat to collaborate with the developers of Smart Autotune Thermostat, they exactly complement in this way and both projects seem under active and increased development.

**Comment by KipK** [comment]
Hi Maia, I've disabled comments on the gist sorry, but it's an internal working document for our wip. 
It is not finished we're only focused for now on the regulation part for electric heaters. 
Valve will comes afterwards.

You can post suggestions in the discussion on my repo.

**Reply by maia** [reply]
> You can post suggestions in the discussion on my repo.
Sure, where do you suggest to place my comment regarding valves, here https://github.com/KipK/versatile_thermostat/discussions/1 or as a new discussion?

**Reply by KipK** [reply]
yes create a new discussion. I'm pretty sure @caiusseverus and @altruizine will have some input on the subject. 

For now we're not working on this as there's priorities to finish first.

**Reply by KipK** [reply]
On another hand, I plan a feature that will auto learn the heating curve itself. 
But as said, I haven't started to think how yet.
